src/plugins/intel_cpu/src/nodes/matmul.cpp

```diff
@@ -110,8 +110,75 @@ bool MatMul::isSupportedOperation(const std::shared_ptr<const ngraph::Node>& op,
     return true;
 }
 
+class MMShapeInfer : public ShapeInferEmptyPads {
+public:
+    MMShapeInfer(const std::shared_ptr<const ngraph::opset1::MatMul>& op) : m_matmul(op) {
+        m_out_rank = m_matmul->get_output_partial_shape(0).rank().get_length();
+        m_YShape = VectorDims(m_out_rank, 1); // for output and cache
+    }
+    std::vector<VectorDims> infer(
+        const std::vector<std::reference_wrapper<const VectorDims>>& input_shapes,
+        const std::unordered_map<size_t, MemoryPtr>& data_dependency) override {
+        VectorDims AShape = input_shapes[0].get();
+        VectorDims BShape = input_shapes[1].get();
+        const size_t& Aranks = AShape.size();
+        const size_t& Branks = BShape.size();
+
+        // getSupportedDescriptors has done some shape check.
+        // 1. Needn't assert the scalar type since the matmul_shape_inference has checked.
+        // 2. Needn't check the compatibility of the last two dims
+        // 3. 1-D x 1-D is needed
+        // 4. transpose is necessary
+        // 5. Just support the same rank of matmul
+        // 6. simplify the broadcast check
+        if (Aranks == 1 && Branks == 1 && AShape[0] == BShape[0]) {
+            return {m_YShape};
+        }
+
+        // transpose
+        const bool& transpose_a = m_matmul->get_transpose_a();
+        const bool& transpose_b = m_matmul->get_transpose_b();
+
+        m_YShape.at(m_out_rank-2) = transpose_a ? AShape[Aranks-1] : AShape[Aranks-2];
```

Comment:
Regarding `at` call. It is more expensive than `[]` since it contains range check. I suggest checking the sizes once at the beginning, that should be enough.

Reply:
> 

Yes. Indeed. at is safe but slow. replace at by []